### PR TITLE
Add e2e for upgrade

### PR DESCRIPTION
**Pull request by nshankar13**
Added a new pipeline-job to test the chart after upgrading from the previous version. 

Note: I'll change the add-extension and update-extension scripts to use k8s-extension as opposed to az rest in a separate PR. 

Note: This PR needs to be merged to main first before I'm actually able connect it to the OSM project on ADO and run the release-e2e-job.

### CODE REVIEW

File: scripts/update-extension.sh [hunk 1]
```diff
@@ -3,22 +3,21 @@
...
 
-export RESOURCEID=subscriptions/$SUBSCRIPTION/resourceGroups/$RESOURCEGROUP/providers/Microsoft.Kubernetes/connectedClusters/$CLUSTERNAME
+az account set --subscription=$SUBSCRIPTION > /dev/null 2>&1
+
+az extension remove --name k8s-extension
+
+git clone https://github.com/Azure/azure-arc-kubernetes-preview
 
-echo $RESOURCEID
+az extension add --source ./extensions/k8s_extension-$K8S_EXTENSION_VERSION-py2.py3-none-any.whl --yes
```
Comment (ksubrmnn) [inline scripts/update-extension.sh:16]: Can you move lines 12-16 out of this script? This won't work for devs running this script locally

File: .pipelines/update-to-current-version.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  - bash: |
+      ./scripts/update-extension.sh
+    displayName: Add Arc extension to AKS cluster
+    env:
+        SUBSCRIPTION: $(SUBSCRIPTION_ID)
+        RESOURCEGROUP: $(AZURE_CLUSTER_NAME)
+        CLUSTERNAME: $(AZURE_CLUSTER_NAME)
+        CHECKOUT_TAG: $(CHECKOUT_TAG)
+        RELEASE_NAMESPACE: $(release.namespace)
+        K8S_EXTENSION_VERSION: $(k8s_extension.version)
```
Comment (ksubrmnn) [inline .pipelines/update-to-current-version.yaml:11]: Missing newline

File: .pipelines/templates/get-latest-tag.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+steps: 
+  - bash: |
+      DESCRIBE_TAG=$(git describe)
+      echo "##vso[task.setvariable variable=LATEST_TAG]${LATEST_TAG:1}"
+      echo ${LATEST_TAG}
+    displayName: Get most recent osm-arc tag
```
Comment (ksubrmnn) [inline .pipelines/templates/get-latest-tag.yaml:6]: Missing newline

**Review by ksubrmnn — CHANGES_REQUESTED** [review]

File: scripts/update-extension.sh (hunk 1 above)
Comment (nshankar13) [reply scripts/update-extension.sh:16]: Good catch, I can move this to be inside the pipeline.

Comment (nshankar13) [reply scripts/update-extension.sh:16]: Users running this locally would have to have k8s-extension already installed before running the script.

File: .pipelines/templates/install-k8s-extension.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+steps: 
+  - bash: |
+      az extension remove --name k8s-extension
+      git clone https://github.com/Azure/azure-arc-kubernetes-preview
+      az extension add --source ./extensions/k8s_extension-${k8s_extension.version}-py2.py3-none-any.whl --yes
```
Comment (ksubrmnn) [inline .pipelines/templates/install-k8s-extension.yaml:5]: is this path correct for the source file? Don't you have to be in the azure-arc-kubernetes-preview directory first?

**Review by ksubrmnn — CHANGES_REQUESTED** [review]

File: .pipelines/templates/install-k8s-extension.yaml (hunk 4 above)
Comment (nshankar13) [reply .pipelines/templates/install-k8s-extension.yaml:5]: Good catch, I also need to add `$(System.DefaultWorkingDirectory)/azure-arc-kubernetes-preview` when I run `git clone`

File: .pipelines/templates/aks-setup.yaml [hunk 5]
```diff
@@ -5,7 +5,7 @@ steps:
     displayName: Create azure cluster name variable
   - script: |
       az login --identity > /dev/null 2>&1
-
+
```
Comment (ksubrmnn) [inline .pipelines/templates/aks-setup.yaml:8]: can you remove this whitespace?

File: .pipelines/templates/package-helm-chart.yaml [hunk 6]
```diff
@@ -1,10 +1,20 @@
...
+    default: false
+
 steps: 
 - script: |
     echo "##vso[task.setvariable variable=UNIQUE_TAG]$(git rev-parse --short HEAD)"
     echo ${UNIQUE_TAG}
   displayName: Create Unique Tag
 - script: |
-    echo "##vso[task.setvariable variable=CHECKOUT_TAG]$(image.tag)-${UNIQUE_TAG}"
+    echo ${{ parameters.releaseTag }}
```
Comment (ksubrmnn) [inline .pipelines/templates/package-helm-chart.yaml:12]: What is the purpose of this change? Do you ever set releaseTag to false? Can we add a comment here explaining this

**Review by ksubrmnn — CHANGES_REQUESTED** [review]

**Comment by ritazh** [comment]
https://github.com/Azure/osm-azure/pull/66#issuecomment-786355997 is related to the upstream e2e test failures

File: .pipelines/release-e2e-job.yaml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    exclude:
+      - docs/*
+
+variables:
+  chart.name: osm-arc
+  chart.path: $(System.DefaultWorkingDirectory)/charts/$(chart.name)
+  image.dir: $(Build.ArtifactStagingDirectory)
+  image.artifact.name: drop
+  KUBECONFIG: $(System.DefaultWorkingDirectory)/kubeconfig.json
+  image.tag: 0.1.0
```
[the remark this answers is not in the document]
Comment (nshankar13) [reply .pipelines/release-e2e-job.yaml:23]: We use image tags of 0.1.0 for all the charts packaged/pushed to staging ACR in the CI.

File: test/bats/test.bats [hunk 8]
```diff
@@ -8,11 +8,25 @@ WAIT_TIME_DEPLOYMENTS=600
...
+    assert_success
+}
+
 @test "arc-osm-system deployments have succeeded" {
     run wait_for_process $WAIT_TIME_DEPLOYMENTS $SLEEP_TIME_DEPLOYMENTS "kubectl wait --for=condition=available deployment --all --namespace arc-osm-system"
     assert_success
 }
 
+@test "only one osm-controller pod" {
+    run wait_for_condition $WAIT_TIME $SLEEP_TIME "$(kubectl get pods -n arc-osm-system -l app=osm-controller | wc -l) == 2"
```
[the remark this answers is not in the document]
Comment (nshankar13) [reply test/bats/test.bats:26]: I think this is the same as what I already have. The osm-controller label is specific to the osm-controller pod.

File: test/bats/test.bats [hunk 9]
```diff
@@ -8,11 +8,25 @@ WAIT_TIME_DEPLOYMENTS=600
...
+    assert_success
```
[the remark this answers is not in the document]
Comment (nshankar13) [reply test/bats/test.bats:27]: https://github.com/Azure/osm-azure/issues/69

File: .pipelines/templates/package-helm-chart.yaml (hunk 6 above)
Comment (nshankar13) [reply .pipelines/templates/package-helm-chart.yaml:12]: The pr e2e will use the default false value of releaseTag, so the packaged chart name will just be osm-arc-<git-hash>. The release-e2e-job will pass in releaseTag as true when it calls this yaml, which adds an additional `-release` to the name of the packaged chart. Otherwise both the release-chart and pr-chart will have identical names.

File: .pipelines/templates/run-upstream-e2e.yaml [hunk 10]
```diff
@@ -10,3 +14,6 @@ steps:
       CTR_REGISTRY: $(UPSTREAM_DOCKER_REGISTRY)
       CTR_TAG: $(UPSTREAM_BRANCH)
     workingDirectory: $(System.DefaultWorkingDirectory)/osm
+  - bash: |
+        kubectl get pods -A
+    displayName: "Get pods to debug"
```
[the remark this answers is not in the document]
Comment (nshankar13) [reply .pipelines/templates/run-upstream-e2e.yaml:19]: I can actually remove this because I'm adding debug steps in a separate PR.

File: test/bats/test.bats (hunk 8 above)
[the remark this answers is not in the document]
Comment (ksubrmnn) [reply test/bats/test.bats:26]: This is not the same because you are just checking for any pods in the osm-controller namespace. You should check specifically for the osm-controller pod, again because the new release introduces an osm-injector pod in the same namespace

File: .pipelines/release-e2e-job.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (ksubrmnn) [reply .pipelines/release-e2e-job.yaml:23]: Why is 0.1.0 used for each chart pushed to staging?

[the remark this answers is not in the document]
Comment (nshankar13) [reply .pipelines/release-e2e-job.yaml:23]: My thinking was that it would be easier to identify which charts we use for just for testing in staging (ex: in case some of them don't get deleted properly). Another reason is that, I don't know if there's a way for the pipeline to know the exact version of the current helm chart specified in Chart.yaml (a string parse with Helm show might work)?. The only other way I can think of to get the current version is to have a variable in the UI that we would continuously update after reach release to reflect the current chart version in main - but we would have to continuously update this after every release.

**Comment by ritazh** [comment]
Thanks for the PR @nshankar13!

One high-level feedback, the upgrade e2e job should be renamed to `upgrade-e2e-job.yaml` as the focus is on upgrade, not release. It should install the last production release in an arc enabled cluster, then upgrade to a release based on PR or main, then run e2es. 

Right now this PR is trying to reuse e2e-aks-arc.yaml. IMO, i think it's better to keep the PR gate e2e logic separate from this upgrade job. You can still reuse/call individual steps and scripts.

**Comment by nshankar13** [comment]
@ritazh I agree that renaming the job/yaml to upgrade-e2e makes sense. 

The reason I decided to reuse the e2e-aks-arc.yaml is that I noticed for other pipelines on ADO, most of them tend to reuse yamls as much as possible by passing in different parameters, as opposed to having two similar, separate yamls. But maybe in this case because the two jobs are different we should use a separate yaml file. @sozercan what do you think?

File: .pipelines/templates/package-helm-chart.yaml (hunk 6 above)
Comment (nshankar13) [reply .pipelines/templates/package-helm-chart.yaml]: Update: I changed the package/publish steps so that publish is able to use the current version of OSM-arc and add a unique id to the image tag before pushing it to staging. For the release-job, we add "-release" to the image name, and for the "pr-job", we add "-pr" to the image name.

File: .pipelines/release-e2e-job.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (nshankar13) [reply .pipelines/release-e2e-job.yaml]: Update: I found a way to use the current version of the chart instead of hardcoding 0.1.0.

File: .pipelines/e2e-aks-arc.yaml [hunk 11]
```diff
@@ -8,15 +13,35 @@ jobs:
     - template: templates/install-oras.yaml
     - template: templates/install-bats.yaml
     - template: templates/package-helm-chart.yaml
-    - template: templates/publish-helm-chart.yaml
+    - ${{ if eq(parameters.releaseE2E, true) }}:
```
Comment (ksubrmnn) [inline .pipelines/e2e-aks-arc.yaml:16]: Can you collapse this into one line where you pass through parameters.releaseE2E for releaseTag?

File: .pipelines/e2e-aks-arc.yaml [hunk 12]
```diff
@@ -8,15 +13,35 @@ jobs:
...
+      - template: templates/publish-helm-chart.yaml
+        parameters:
+          releaseTag: true
+    - ${{ if eq(parameters.releaseE2E, false) }}:
+      - template: templates/publish-helm-chart.yaml
+    - template: templates/checkout-upstream-repo.yaml
     - template: templates/aks-setup.yaml
     - template: templates/enable-arc.yaml
-    - template: templates/add-arc-extension.yaml
+    - ${{ if eq(parameters.releaseE2E, true) }}:
```
Comment (ksubrmnn) [inline .pipelines/e2e-aks-arc.yaml:25]: Can you separate this part (lines 26-37) out into a separate file? For readability

File: .pipelines/templates/checkout-upstream-repo.yaml [hunk 13]
```diff
@@ -0,0 +1,4 @@
+steps:
+  - bash: |
+      git clone -b v${CHART_TAG} $(upstream.repo) $(System.DefaultWorkingDirectory)/osm
```
Comment (ksubrmnn) [inline .pipelines/templates/checkout-upstream-repo.yaml:3]: Should this be `$(System.DefaultWorkingDirectory)/openservicemesh` to match the repo?

File: .pipelines/templates/add-arc-extension.yaml [hunk 14]
```diff
@@ -1,11 +1,17 @@
+parameters: 
+  - name: imageTag
+    type: string
+    default: $(CHECKOUT_TAG)
+
 steps:  
   - bash: |
+      echo ${{ parameters.imageTag }}
```
Comment (ksubrmnn) [inline .pipelines/templates/add-arc-extension.yaml:8]: Can you add more context to this log message. Something like "Image tag: $parameters.imageTag"

File: .pipelines/templates/publish-helm-chart.yaml [hunk 15]
```diff
@@ -1,15 +1,41 @@
...
     displayName: Login to acr
     env: 
       REGISTRY_NAME: $(REGISTRY_NAME)
+  - script: |
+      cp $(image.dir)/$(chart.name)-*.tgz .
+      FILENAME=$(ls $(chart.name)-*.tgz)
+      TRIM=${FILENAME##$(chart.name)-}
+      echo "##vso[task.setvariable variable=CHART_TAG]${TRIM%%.tgz}"
+      echo ${CHART_TAG}
+    displayName: Get osm-arc chart version
```
Comment (ksubrmnn) [inline .pipelines/templates/publish-helm-chart.yaml:19]: Can you make this more descriptive? Like "get osm-arc chart version from built and published chart"

File: .pipelines/templates/publish-helm-chart.yaml [hunk 16]
```diff
@@ -1,15 +1,41 @@
...
+      echo ${CHECKOUT_TAG}
+    displayName: Create Checkout Tag 
   - bash: |
       if [[ -n "${CHECKOUT_TAG}" ]]; then
-        cp $(image.dir)/$(chart.name)-${CHECKOUT_TAG}.tgz .
         echo 'pushing chart'
-        oras push $(REPOSITORY_NAME):${CHECKOUT_TAG} ./$(chart.name)-${CHECKOUT_TAG}.tgz:application/tar+gzip --debug
+        oras push $(REPOSITORY_NAME):${CHECKOUT_TAG} ./$(chart.name)-${CHART_TAG}.tgz:application/tar+gzip --debug
       else 
         echo "Helm chart was not published to staging ACR because $CHECKOUT_TAG was not set by the pipeline" && exit 1
```
Comment (ksubrmnn) [inline .pipelines/templates/publish-helm-chart.yaml:40]: Can you rename this to say CHART_TAG instead of CHECKOUT_TAG

**Review by ksubrmnn — CHANGES_REQUESTED** [review]
Have some small changes/questions. Let's get this merged without too much refactoring and we can go back in follow ups to improve structure/readability. This way you can test in the interim.

File: .pipelines/e2e-aks-arc.yaml (hunk 11 above)
Comment (nshankar13) [reply .pipelines/e2e-aks-arc.yaml:16]: good catch

File: .pipelines/templates/checkout-upstream-repo.yaml (hunk 13 above)
Comment (nshankar13) [reply .pipelines/templates/checkout-upstream-repo.yaml:3]: https://github.com/Azure/osm-azure/pull/38#discussion_r568896100

File: .pipelines/templates/publish-helm-chart.yaml (hunk 16 above)
Comment (nshankar13) [reply .pipelines/templates/publish-helm-chart.yaml]: Need to use CHECKOUT_TAG for the image tag that gets pushed to staging because CHECKOUT_TAG contains the unique identifier for the chart in each PR. I will rename all the tag variables in a separate PR to make these variable names more intuitive.

File: .pipelines/upgrade-templates.yaml [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+    parameters: 
+      imageTag: $(LATEST_TAG)
+  - script: |
+      make test-e2e
+    displayName: "Run osm-azure e2e tests on latest release chart"
+    env: 
+      EXTENSION_TAG: $(LATEST_TAG)
+      KUBECONFIG: $(System.DefaultWorkingDirectory)/kubeconfig.json
+  - template: templates/delete-CRDS.yaml
+  - template: templates/update-arc-addon.yaml
```
Comment (ksubrmnn) [inline .pipelines/upgrade-templates.yaml:13]: add a newline

File: .pipelines/upgrade-templates.yaml [hunk 18]
```diff
@@ -0,0 +1,13 @@
+steps:    
+  - template: templates/get-latest-tag.yaml
```
Comment (ksubrmnn) [inline .pipelines/upgrade-templates.yaml:2]: Add comment to make this clear that latest is the previously published tag

File: .pipelines/templates/delete-CRDS.yaml [hunk 19]
```diff
@@ -0,0 +1,11 @@
+steps:
```
Comment (ksubrmnn) [inline .pipelines/templates/delete-CRDS.yaml:1]: Rename to upgrade CRDS

**Review by ksubrmnn — APPROVED** [review]
I'm good with merging this for testing and you can address nits in a follow up

**Review by ksubrmnn — APPROVED** [review]

File: .pipelines/upgrade-templates.yaml (hunk 18 above)
Comment (nshankar13) [reply .pipelines/upgrade-templates.yaml:2]: Will do this in a follow up PR

**Review by draychev — APPROVED** [review]

File: .pipelines/upgrade-templates.yaml (hunk 18 above)
Comment (nshankar13) [reply .pipelines/upgrade-templates.yaml:2]: As a part of: https://github.com/Azure/osm-azure/issues/75
